File: packages/nexus-reactor/src/parser/parser.ts
/**
 * @nexus/reactor - NXML Parser
 * 
 * Parses NXML tokens into a NexusPanelAST structure.
 */

import type { Token, TokenType } from './lexer';
import { tokenize, getTokenTypeName } from './lexer';
import { ParseError } from '../core/errors';
import type {
  NexusPanelAST,
  PanelMeta,
  DataAST,
  LogicAST,
  ViewAST,
  StateNode,
  ComputedNode,
  ToolNode,
  ArgNode,
  HandlerNode,
  LifecycleNode,
  ExtensionNode,
  ViewNode,
  SourceLocation,
  NXMLPrimitiveType,
  LifecycleEvent,
} from '../core/types';
import { parseDefaultValue } from '../utils/coercion';

interface ParserState {
  tokens: Token[];
  current: number;
}

export function parse(source: string): NexusPanelAST {
  const tokens = tokenize(source);
  const state: ParserState = { tokens, current: 0 };
  return parseNexusPanel(state);
}

function parseNexusPanel(state: ParserState): NexusPanelAST {
  expect(state, 'TAG_OPEN');
  const tagName = expect(state, 'TAG_NAME');
  
  if (tagName.value !== 'NexusPanel') {
    throw new ParseError(`Root element must be <NexusPanel>, got <${tagName.value}>`, { loc: tagName.loc });
  }

  const meta = parsePanelMeta(state);
  const closeToken = peek(state);
  if (closeToken.type === 'TAG_SELF_CLOSE') {
    throw new ParseError('NexusPanel cannot be self-closing', { loc: closeToken.loc });
  }
  expect(state, 'TAG_CLOSE');

  let data: DataAST | undefined;
  let logic: LogicAST | undefined;
  let view: ViewAST | undefined;

  while (!isAtEnd(state) && !check(state, 'TAG_END_OPEN')) {
    if (check(state, 'TAG_OPEN')) {
      advance(state);
      const childTag = expect(state, 'TAG_NAME');
      
      switch (childTag.value) {
        case 'Data':
          if (data) throw new ParseError('Duplicate <Data> section', { loc: childTag.loc });
          data = parseDataSection(state, childTag.loc);
          break;
        case 'Logic':
          if (logic) throw new ParseError('Duplicate <Logic> section', { loc: childTag.loc });
          logic = parseLogicSection(state, childTag.loc);
          break;
        case 'View':
          if (view) throw new ParseError('Duplicate <View> section', { loc: childTag.loc });
          view = parseViewSection(state, childTag.loc);
          break;
        default:
          throw new ParseError(`Unknown section <${childTag.value}> in NexusPanel`, { loc: childTag.loc });
      }
    } else {
      advance(state);
    }
  }

  expect(state, 'TAG_END_OPEN');
  const endTagName = expect(state, 'TAG_NAME');
  if (endTagName.value !== 'NexusPanel') {
    throw new ParseError(`Expected </NexusPanel>, got </${endTagName.value}>`, { loc: endTagName.loc });
  }
  expect(state, 'TAG_CLOSE');

  return {
    kind: 'NexusPanel',
    meta,
    data: data ?? { kind: 'Data', states: [], computed: [] },
    logic: logic ?? { kind: 'Logic', extensions: [], tools: [], lifecycles: [] },
    view: view ?? { kind: 'View', root: { type: 'Layout', props: {}, children: [] } },
  };
}

function parsePanelMeta(state: ParserState): PanelMeta {
  const attrs = parseAttributes(state);
  return {
    title: attrs.title as string ?? 'Untitled Panel',
    description: attrs.description as string,
    id: attrs.id as string,
    version: attrs.version as string,
    author: attrs.author as string,
    tags: attrs.tags ? String(attrs.tags).split(',').map(t => t.trim()) : undefined,
  };
}

function parseDataSection(state: ParserState, startLoc: SourceLocation): DataAST {
  const states: StateNode[] = [];
  const computed: ComputedNode[] = [];

  skipAttributes(state);
  if (check(state, 'TAG_SELF_CLOSE')) {
    advance(state);
    return { kind: 'Data', states, computed, loc: startLoc };
  }
  expect(state, 'TAG_CLOSE');

  while (!isAtEnd(state) && !check(state, 'TAG_END_OPEN')) {
    if (check(state, 'TAG_OPEN')) {
      advance(state);
      const childTag = expect(state, 'TAG_NAME');
      switch (childTag.value) {
        case 'State':
          states.push(parseStateNode(state, childTag.loc));
          break;
        case 'Computed':
          computed.push(parseComputedNode(state, childTag.loc));
          break;
        default:
          throw new ParseError(`Unknown element <${childTag.value}> in Data section`, { loc: childTag.loc });
      }
    } else {
      advance(state);
    }
  }
  expectClosingTag(state, 'Data');
  return { kind: 'Data', states, computed, loc: startLoc };
}

function parseStateNode(state: ParserState, loc: SourceLocation): StateNode {
  const attrs = parseAttributes(state);
  expectSelfCloseOrEmpty(state, 'State');
  const name = attrs.name as string;
  const type = attrs.type as NXMLPrimitiveType ?? 'string';
  const defaultValue = parseDefaultValue(attrs.default as string, type);
  if (!name) throw new ParseError('State must have a name attribute', { loc });
  return { kind: 'State', name, type, default: defaultValue, loc };
}

function parseComputedNode(state: ParserState, loc: SourceLocation): ComputedNode {
  const attrs = parseAttributes(state);
  expectSelfCloseOrEmpty(state, 'Computed');
  const name = attrs.name as string;
  const value = attrs.value as string;
  if (!name) throw new ParseError('Computed must have a name attribute', { loc });
  if (!value) throw new ParseError('Computed must have a value attribute', { loc });
  return { kind: 'Computed', name, value, loc };
}

function parseLogicSection(state: ParserState, startLoc: SourceLocation): LogicAST {
  const extensions: ExtensionNode[] = [];
  const tools: ToolNode[] = [];
  const lifecycles: LifecycleNode[] = [];

  skipAttributes(state);
  if (check(state, 'TAG_SELF_CLOSE')) {
    advance(state);
    return { kind: 'Logic', extensions, tools, lifecycles, loc: startLoc };
  }
  expect(state, 'TAG_CLOSE');

  while (!isAtEnd(state) && !check(state, 'TAG_END_OPEN')) {
    if (check(state, 'TAG_OPEN')) {
      advance(state);
      const childTag = expect(state, 'TAG_NAME');
      switch (childTag.value) {
        case 'Extension':
          extensions.push(parseExtensionNode(state, childTag.loc));
          break;
        case 'Tool':
          tools.push(parseToolNode(state, childTag.loc));
          break;
        case 'Lifecycle':
          lifecycles.push(parseLifecycleNode(state, childTag.loc));
          break;
        default:
          throw new ParseError(`Unknown element <${childTag.value}> in Logic section`, { loc: childTag.loc });
      }
    } else {
      advance(state);
    }
  }
  expectClosingTag(state, 'Logic');
  return { kind: 'Logic', extensions, tools, lifecycles, loc: startLoc };
}

function parseExtensionNode(state: ParserState, loc: SourceLocation): ExtensionNode {
  const attrs = parseAttributes(state);
  expectSelfCloseOrEmpty(state, 'Extension');
  const name = attrs.name as string;
  const alias = (attrs.alias as string) ?? name.split('.').pop() ?? name;
  const source = attrs.source as string;
  if (!name) throw new ParseError('Extension must have a name attribute', { loc });
  return { kind: 'Extension', name, alias, source, loc };
}

function parseToolNode(state: ParserState, loc: SourceLocation): ToolNode {
  const attrs = parseAttributes(state);
  const name = attrs.name as string;
  const description = attrs.description as string;
  if (!name) throw new ParseError('Tool must have a name attribute', { loc });

  if (check(state, 'TAG_SELF_CLOSE')) {
    advance(state);
    throw new ParseError('Tool must have a Handler child', { loc });
  }
  expect(state, 'TAG_CLOSE');

  const args: ArgNode[] = [];
  let handler: HandlerNode | undefined;

  while (!isAtEnd(state) && !check(state, 'TAG_END_OPEN')) {
    if (check(state, 'TAG_OPEN')) {
      advance(state);
      const childTag = expect(state, 'TAG_NAME');
      switch (childTag.value) {
        case 'Arg':
          args.push(parseArgNode(state, childTag.loc));
          break;
        case 'Handler':
          handler = parseHandlerNode(state, childTag.loc);
          break;
        default:
          throw new ParseError(`Unknown element <${childTag.value}> in Tool`, { loc: childTag.loc });
      }
    } else {
      advance(state);
    }
  }

  if (!handler) throw new ParseError('Tool must have a Handler child', { loc });
  expectClosingTag(state, 'Tool');
  return { kind: 'Tool', name, description, args, handler, loc };
}

function parseArgNode(state: ParserState, loc: SourceLocation): ArgNode {
  const attrs = parseAttributes(state);
  expectSelfCloseOrEmpty(state, 'Arg');
  const name = attrs.name as string;
  const type = attrs.type as NXMLPrimitiveType ?? 'string';
  const required = attrs.required !== 'false';
  const defaultValue = attrs.default;
  const description = attrs.description as string;
  if (!name) throw new ParseError('Arg must have a name attribute', { loc });
  return { kind: 'Arg', name, type, required, default: defaultValue, description, loc };
}

function parseHandlerNode(state: ParserState, loc: SourceLocation): HandlerNode {
  skipAttributes(state);
  if (check(state, 'TAG_SELF_CLOSE')) {
    advance(state);
    return { kind: 'Handler', code: '', loc };
  }
  expect(state, 'TAG_CLOSE');

  let code = '';
  while (!isAtEnd(state) && !check(state, 'TAG_END_OPEN')) {
    const token = peek(state);
    if (token.type === 'TEXT' || token.type === 'CDATA') {
      code += token.value;
    }
    advance(state);
  }
  expectClosingTag(state, 'Handler');
  const isAsync = /\bawait\b/.test(code);
  return { kind: 'Handler', code: code.trim(), isAsync, loc };
}

function parseLifecycleNode(state: ParserState, loc: SourceLocation): LifecycleNode {
  const attrs = parseAttributes(state);
  const on = attrs.on as LifecycleEvent;
  if (on !== 'mount' && on !== 'unmount') {
    throw new ParseError(`Lifecycle "on" attribute must be "mount" or "unmount", got "${on}"`, { loc });
  }

  if (check(state, 'TAG_SELF_CLOSE')) {
    advance(state);
    throw new ParseError('Lifecycle must have a Handler child', { loc });
  }
  expect(state, 'TAG_CLOSE');

  let handler: HandlerNode | undefined;
  while (!isAtEnd(state) && !check(state, 'TAG_END_OPEN')) {
    if (check(state, 'TAG_OPEN')) {
      advance(state);
      const childTag = expect(state, 'TAG_NAME');
      if (childTag.value === 'Handler') {
        handler = parseHandlerNode(state, childTag.loc);
      } else {
        throw new ParseError(`Lifecycle can only contain Handler, got <${childTag.value}>`, { loc: childTag.loc });
      }
    } else {
      advance(state);
    }
  }
  if (!handler) throw new ParseError('Lifecycle must have a Handler child', { loc });
  expectClosingTag(state, 'Lifecycle');
  return { kind: 'Lifecycle', on, handler, loc };
}

function parseViewSection(state: ParserState, startLoc: SourceLocation): ViewAST {
  skipAttributes(state);
  if (check(state, 'TAG_SELF_CLOSE')) {
    advance(state);
    return { kind: 'View', root: { type: 'Layout', props: {}, children: [] }, loc: startLoc };
  }
  expect(state, 'TAG_CLOSE');

  let root: ViewNode | undefined;
  while (!isAtEnd(state) && !check(state, 'TAG_END_OPEN')) {
    if (check(state, 'TAG_OPEN')) {
      if (root) throw new ParseError('View can only have one root element', { loc: peek(state).loc });
      root = parseViewNode(state);
    } else {
      advance(state);
    }
  }
  expectClosingTag(state, 'View');
  return { kind: 'View', root: root ?? { type: 'Layout', props: {}, children: [] }, loc: startLoc };
}

function parseViewNode(state: ParserState): ViewNode {
  expect(state, 'TAG_OPEN');
  const tagToken = expect(state, 'TAG_NAME');
  const tagName = tagToken.value;
  const attrs = parseAttributes(state);

  // Special handling for CustomComponent with bind: and on: prefixes
  let props: Record<string, unknown>;
  let id: string | undefined;

  if (tagName === 'CustomComponent') {
    const bindings: Record<string, string> = {};
    const events: Record<string, string> = {};
    const regularProps: Record<string, unknown> = {};

    for (const [key, value] of Object.entries(attrs)) {
      if (key.startsWith('bind:')) {
        // bind:document="$state.document" -> bindings.document = "$state.document"
        const propName = key.slice(5); // Remove 'bind:' prefix
        bindings[propName] = value;
      } else if (key.startsWith('on:')) {
        // on:change="(doc) => { ... }" -> events.change = "(doc) => { ... }"
        const eventName = key.slice(3); // Remove 'on:' prefix
        events[eventName] = value;
      } else {
        regularProps[key] = value;
      }
    }

    id = regularProps.id as string | undefined;
    props = {
      ...regularProps,
      ...(Object.keys(bindings).length > 0 && { bindings }),
      ...(Object.keys(events).length > 0 && { events }),
    };
  } else {
    // Standard component handling
    props = { ...attrs };
    id = props.id as string | undefined;
  }

  if (check(state, 'TAG_SELF_CLOSE')) {
    advance(state);
    return { type: tagName, id, props, children: [], loc: tagToken.loc };
  }
  expect(state, 'TAG_CLOSE');

  const children: ViewNode[] = [];
  while (!isAtEnd(state) && !check(state, 'TAG_END_OPEN')) {
    if (check(state, 'TAG_OPEN')) {
      children.push(parseViewNode(state));
    } else {
      advance(state);
    }
  }
  expectClosingTag(state, tagName);
  return { type: tagName, id, props, children, loc: tagToken.loc };
}

// Helper functions
function parseAttributes(state: ParserState): Record<string, string> {
  const attrs: Record<string, string> = {};
  while (check(state, 'ATTR_NAME')) {
    const nameToken = advance(state);
    if (check(state, 'EQUALS')) {
      advance(state);
      if (check(state, 'ATTR_VALUE')) {
        const valueToken = advance(state);
        let value = valueToken.value;
        if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
          value = value.slice(1, -1);
        }
        attrs[nameToken.value] = value;
      }
    } else {
      attrs[nameToken.value] = 'true';
    }
  }
  return attrs;
}

function skipAttributes(state: ParserState): void {
  while (check(state, 'ATTR_NAME')) {
    advance(state);
    if (check(state, 'EQUALS')) {
      advance(state);
      if (check(state, 'ATTR_VALUE')) advance(state);
    }
  }
}

function expectClosingTag(state: ParserState, tagName: string): void {
  expect(state, 'TAG_END_OPEN');
  const endTag = expect(state, 'TAG_NAME');
  if (endTag.value !== tagName) {
    throw new ParseError(`Expected </${tagName}>, got </${endTag.value}>`, { loc: endTag.loc });
  }
  expect(state, 'TAG_CLOSE');
}

function expectSelfCloseOrEmpty(state: ParserState, tagName: string): void {
  if (check(state, 'TAG_SELF_CLOSE')) {
    advance(state);
  } else if (check(state, 'TAG_CLOSE')) {
    advance(state);
    expectClosingTag(state, tagName);
  }
}

function peek(state: ParserState): Token {
  return state.tokens[state.current] ?? { type: 'EOF', value: '', loc: { startLine: 0, startColumn: 0, endLine: 0, endColumn: 0 } };
}

function check(state: ParserState, type: TokenType): boolean {
  return peek(state).type === type;
}

function advance(state: ParserState): Token {
  if (!isAtEnd(state)) state.current++;
  return state.tokens[state.current - 1];
}

function expect(state: ParserState, type: TokenType): Token {
  const token = peek(state);
  if (token.type !== type) {
    throw ParseError.unexpectedToken(token.value || getTokenTypeName(token.type), getTokenTypeName(type), token.loc);
  }
  return advance(state);
}

function isAtEnd(state: ParserState): boolean {
  return peek(state).type === 'EOF';
}

File: runtime/nexus-core/nexus_core/parser/parser.py
"""
NXML Parser - Builds AST from tokens.
"""

from typing import List, Dict, Any, Optional
from nexus_protocol.ast import (
    NexusPanelAST,
    PanelMeta,
    DataAST,
    LogicAST,
    ViewAST,
    StateNode,
    ComputedNode,
    ToolNode,
    HandlerNode,
    LifecycleNode,
    ExtensionNode,
    ArgNode,
    ViewNode,
    SourceLocation,
    NXMLPrimitiveType,
)
from .lexer import Lexer, Token, TokenType
from datetime import datetime


class ParseError(Exception):
    """Parse error with location."""

    def __init__(self, message: str, token: Optional[Token] = None):
        self.message = message
        self.token = token
        if token:
            super().__init__(
                f"{message} at line {token.loc.start_line}, column {token.loc.start_column}"
            )
        else:
            super().__init__(message)


class Parser:
    """NXML parser that builds Pydantic AST from tokens."""

    def __init__(self, source: str):
        self.source = source
        self.lexer = Lexer(source)
        self.tokens = self.lexer.tokenize()
        self.current = 0

    def parse(self) -> NexusPanelAST:
        """Parse NXML into AST."""
        return self._parse_nexus_panel()

    def _parse_nexus_panel(self) -> NexusPanelAST:
        """Parse root <NexusPanel> element."""
        # Expect <NexusPanel>
        self._expect(TokenType.TAG_OPEN)
        tag_name_token = self._expect(TokenType.TAG_NAME)

        if tag_name_token.value != "NexusPanel":
            raise ParseError(
                f"Root element must be <NexusPanel>, got <{tag_name_token.value}>",
                tag_name_token,
            )

        # Parse attributes (panel metadata)
        meta = self._parse_panel_meta()

        # Expect >
        self._expect(TokenType.TAG_CLOSE)

        # Parse sections
        data: Optional[DataAST] = None
        logic: Optional[LogicAST] = None
        view: Optional[ViewAST] = None

        while not self._check(TokenType.TAG_END_OPEN):
            section_name = self._peek_section_name()

            if section_name == "Data":
                data = self._parse_data_section()
            elif section_name == "Logic":
                logic = self._parse_logic_section()
            elif section_name == "View":
                view = self._parse_view_section()
            else:
                raise ParseError(f"Unknown section: {section_name}")

        # Expect </NexusPanel>
        self._expect(TokenType.TAG_END_OPEN)
        closing_tag = self._expect(TokenType.TAG_NAME)
        if closing_tag.value != "NexusPanel":
            raise ParseError(f"Expected </NexusPanel>, got </{closing_tag.value}>")
        self._expect(TokenType.TAG_CLOSE)

        # Build AST with defaults for missing sections
        return NexusPanelAST(
            kind="NexusPanel",
            meta=meta,
            data=data or DataAST(kind="Data", states=[], computed=[]),
            logic=logic
            or LogicAST(kind="Logic", extensions=[], tools=[], lifecycles=[]),
            view=view
            or ViewAST(
                kind="View",
                root=ViewNode(type="Layout", props={}, children=[]),
            ),
        )

    def _parse_panel_meta(self) -> PanelMeta:
        """Parse panel metadata from attributes."""
        attrs = self._parse_attributes()

        return PanelMeta(
            id=attrs.get("id", ""),
            title=attrs.get("title"),
            type=attrs.get("type", "custom"),
            version=attrs.get("version", "1.0.0"),
            description=attrs.get("description"),
            author=attrs.get("author"),
            tags=attrs.get("tags", "").split(",") if attrs.get("tags") else [],
            icon=attrs.get("icon"),
        )

    def _parse_data_section(self) -> DataAST:
        """Parse <Data> section."""
        self._consume_tag_open("Data")

        states: List[StateNode] = []
        computed: List[ComputedNode] = []

        while not self._check_closing_tag("Data"):
            tag_name = self._peek_section_name()

            if tag_name == "State":
                states.append(self._parse_state_node())
            elif tag_name == "Computed":
                computed.append(self._parse_computed_node())
            else:
                # Skip unknown tags
                self._skip_tag()

        self._consume_tag_close("Data")

        return DataAST(kind="Data", states=states, computed=computed)

    def _parse_state_node(self) -> StateNode:
        """Parse <State> node."""
        self._expect(TokenType.TAG_OPEN)
        self._expect_tag_name("State")

        attrs = self._parse_attributes()

        # Self-closing or has content
        if self._check(TokenType.TAG_SELF_CLOSE):
            self._advance()
        else:
            self._expect(TokenType.TAG_CLOSE)
            # Skip any content
            while not self._check_closing_tag("State"):
                self._advance()
            self._consume_tag_close("State")

        return StateNode(
            kind="State",
            name=attrs.get("name", ""),
            type=NXMLPrimitiveType(attrs.get("type", "any")),
            default=self._parse_default_value(attrs.get("default")),
            description=attrs.get("description"),
        )

    def _parse_computed_node(self) -> ComputedNode:
        """Parse <Computed> node."""
        self._expect(TokenType.TAG_OPEN)
        self._expect_tag_name("Computed")

        attrs = self._parse_attributes()

        if self._check(TokenType.TAG_SELF_CLOSE):
            self._advance()
        else:
            self._expect(TokenType.TAG_CLOSE)
            while not self._check_closing_tag("Computed"):
                self._advance()
            self._consume_tag_close("Computed")

        return ComputedNode(
            kind="Computed",
            name=attrs.get("name", ""),
            value=attrs.get("value", ""),
            type=NXMLPrimitiveType(attrs.get("type", "any"))
            if attrs.get("type")
            else None,
            description=attrs.get("description"),
        )

    def _parse_logic_section(self) -> LogicAST:
        """Parse <Logic> section."""
        self._consume_tag_open("Logic")

        extensions: List[ExtensionNode] = []
        tools: List[ToolNode] = []
        lifecycles: List[LifecycleNode] = []

        while not self._check_closing_tag("Logic"):
            tag_name = self._peek_section_name()

            if tag_name == "Extension":
                extensions.append(self._parse_extension_node())
            elif tag_name == "Tool":
                tools.append(self._parse_tool_node())
            elif tag_name == "Lifecycle":
                lifecycles.append(self._parse_lifecycle_node())
            else:
                self._skip_tag()

        self._consume_tag_close("Logic")

        return LogicAST(
            kind="Logic", extensions=extensions, tools=tools, lifecycles=lifecycles
        )

    def _parse_extension_node(self) -> ExtensionNode:
        """Parse <Extension> node."""
        self._expect(TokenType.TAG_OPEN)
        self._expect_tag_name("Extension")

        attrs = self._parse_attributes()

        if self._check(TokenType.TAG_SELF_CLOSE):
            self._advance()
        else:
            self._expect(TokenType.TAG_CLOSE)
            while not self._check_closing_tag("Extension"):
                self._advance()
            self._consume_tag_close("Extension")

        return ExtensionNode(kind="Extension", name=attrs.get("name", ""), config={})

    def _parse_tool_node(self) -> ToolNode:
        """Parse <Tool> node."""
        self._consume_tag_open("Tool")
        attrs = self._parse_attributes()

        # Parse child elements
        args: List[ArgNode] = []
        handler: Optional[HandlerNode] = None

        while not self._check_closing_tag("Tool"):
            tag_name = self._peek_section_name()

            if tag_name == "Arg":
                args.append(self._parse_arg_node())
            elif tag_name == "Handler":
                handler = self._parse_handler_node()
            else:
                self._skip_tag()

        self._consume_tag_close("Tool")

        if not handler:
            raise ParseError("Tool must have a <Handler>")

        return ToolNode(
            kind="Tool",
            name=attrs.get("name", ""),
            args=args,
            handler=handler,
            description=attrs.get("description"),
            icon=attrs.get("icon"),
        )

    def _parse_arg_node(self) -> ArgNode:
        """Parse <Arg> node."""
        self._expect(TokenType.TAG_OPEN)
        self._expect_tag_name("Arg")

        attrs = self._parse_attributes()

        if self._check(TokenType.TAG_SELF_CLOSE):
            self._advance()
        else:
            self._expect(TokenType.TAG_CLOSE)
            while not self._check_closing_tag("Arg"):
                self._advance()
            self._consume_tag_close("Arg")

        return ArgNode(
            name=attrs.get("name", ""),
            type=NXMLPrimitiveType(attrs.get("type", "any")),
            default=self._parse_default_value(attrs.get("default")),
            required=attrs.get("required", "true").lower() == "true",
            description=attrs.get("description"),
        )

    def _parse_handler_node(self) -> HandlerNode:
        """Parse <Handler> node."""
        self._consume_tag_open("Handler")
        attrs = self._parse_attributes()

        # Get handler code (CODE_BLOCK token)
        code = ""
        if self._check(TokenType.CODE_BLOCK):
            code_token = self._advance()
            code = code_token.value

        self._consume_tag_close("Handler")

        capabilities = (
            attrs.get("capabilities", "").split(",")
            if attrs.get("capabilities")
            else []
        )

        return HandlerNode(
            kind="Handler",
            code=code,
            capabilities=[c.strip() for c in capabilities if c.strip()],
            timeout_ms=int(attrs.get("timeout_ms", "5000")),
        )

    def _parse_lifecycle_node(self) -> LifecycleNode:
        """Parse <Lifecycle> node."""
        self._consume_tag_open("Lifecycle")
        attrs = self._parse_attributes()

        handler: Optional[HandlerNode] = None

        while not self._check_closing_tag("Lifecycle"):
            tag_name = self._peek_section_name()

            if tag_name == "Handler":
                handler = self._parse_handler_node()
            else:
                self._skip_tag()

        self._consume_tag_close("Lifecycle")

        if not handler:
            raise ParseError("Lifecycle must have a <Handler>")

        return LifecycleNode(
            kind="Lifecycle", event=attrs.get("on", ""), handler=handler
        )

    def _parse_view_section(self) -> ViewAST:
        """Parse <View> section."""
        self._consume_tag_open("View")

        # Parse root view node
        root = self._parse_view_node()

        self._consume_tag_close("View")

        return ViewAST(kind="View", root=root)

    def _parse_view_node(self) -> ViewNode:
        """Parse a view component node recursively."""
        self._expect(TokenType.TAG_OPEN)
        tag_name_token = self._expect(TokenType.TAG_NAME)
        component_type = tag_name_token.value

        attrs = self._parse_attributes()

        # Check for self-closing
        if self._check(TokenType.TAG_SELF_CLOSE):
            self._advance()
            return ViewNode(type=component_type, props=attrs, children=[])

        self._expect(TokenType.TAG_CLOSE)

        # Parse children
        children: List[ViewNode] = []
        while not self._check_closing_tag(component_type):
            # Check if next is a tag
            if self._check(TokenType.TAG_OPEN):
                children.append(self._parse_view_node())
            elif self._check(TokenType.TEXT):
                # Text node - add as Text component
                text_token = self._advance()
                children.append(
                    ViewNode(
                        type="Text", props={"value": text_token.value}, children=[]
                    )
                )
            else:
                self._advance()

        self._consume_tag_close(component_type)

        return ViewNode(type=component_type, props=attrs, children=children)

    # Helper methods

    def _parse_attributes(self) -> Dict[str, str]:
        """Parse attributes and return as dict."""
        attrs: Dict[str, str] = {}

        while self._check(TokenType.ATTR_NAME):
            name_token = self._advance()
            name = name_token.value

            if self._check(TokenType.EQUALS):
                self._advance()
                if self._check(TokenType.ATTR_VALUE):
                    value_token = self._advance()
                    attrs[name] = value_token.value
                else:
                    attrs[name] = "true"
            else:
                attrs[name] = "true"

        return attrs

    def _parse_default_value(self, value_str: Optional[str]) -> Any:
        """Parse default value string to appropriate type."""
        if value_str is None:
            return None

        # Try to parse as number
        try:
            if "." in value_str:
                return float(value_str)
            return int(value_str)
        except ValueError:
            pass

        # Check boolean
        if value_str.lower() in ("true", "false"):
            return value_str.lower() == "true"

        # Return as string
        return value_str

    def _consume_tag_open(self, expected_name: str):
        """Consume opening tag."""
        self._expect(TokenType.TAG_OPEN)
        self._expect_tag_name(expected_name)
        self._parse_attributes()  # Parse and discard attributes
        self._expect(TokenType.TAG_CLOSE)

    def _consume_tag_close(self, expected_name: str):
        """Consume closing tag."""
        self._expect(TokenType.TAG_END_OPEN)
        self._expect_tag_name(expected_name)
        self._expect(TokenType.TAG_CLOSE)

    def _skip_tag(self):
        """Skip an entire tag and its contents."""
        if self._check(TokenType.TAG_OPEN):
            self._advance()
            tag_name_token = self._expect(TokenType.TAG_NAME)
            tag_name = tag_name_token.value

            # Skip attributes
            while self._check(TokenType.ATTR_NAME):
                self._advance()
                if self._check(TokenType.EQUALS):
                    self._advance()
                if self._check(TokenType.ATTR_VALUE):
                    self._advance()

            # Self-closing?
            if self._check(TokenType.TAG_SELF_CLOSE):
                self._advance()
                return

            self._expect(TokenType.TAG_CLOSE)

            # Skip content until closing tag
            depth = 1
            while depth > 0 and not self._is_at_end():
                if self._check(TokenType.TAG_OPEN):
                    self._advance()
                    self._advance()  # Tag name
                    depth += 1
                elif self._check(TokenType.TAG_END_OPEN):
                    self._advance()
                    closing_name = self._advance()
                    if closing_name.value == tag_name:
                        depth -= 1
                    self._expect(TokenType.TAG_CLOSE)
                else:
                    self._advance()

    def _peek_section_name(self) -> str:
        """Peek at the next section name."""
        if self._check(TokenType.TAG_OPEN):
            # Look ahead
            next_pos = self.current + 1
            if next_pos < len(self.tokens):
                next_token = self.tokens[next_pos]
                if next_token.type == TokenType.TAG_NAME:
                    return next_token.value
        return ""

    def _check_closing_tag(self, tag_name: str) -> bool:
        """Check if next tokens are a closing tag."""
        if not self._check(TokenType.TAG_END_OPEN):
            return False

        # Look ahead for tag name
        next_pos = self.current + 1
        if next_pos < len(self.tokens):
            next_token = self.tokens[next_pos]
            if next_token.type == TokenType.TAG_NAME and next_token.value == tag_name:
                return True

        return False

    def _expect_tag_name(self, expected: str) -> Token:
        """Expect specific tag name."""
        token = self._expect(TokenType.TAG_NAME)
        if token.value != expected:
            raise ParseError(f"Expected tag name '{expected}', got '{token.value}'", token)
        return token

    def _check(self, token_type: TokenType) -> bool:
        """Check if current token is of given type."""
        if self._is_at_end():
            return False
        return self._peek().type == token_type

    def _expect(self, token_type: TokenType) -> Token:
        """Expect token of specific type."""
        if not self._check(token_type):
            current = self._peek()
            raise ParseError(
                f"Expected {token_type.value}, got {current.type.value}", current
            )
        return self._advance()

    def _advance(self) -> Token:
        """Advance to next token."""
        if not self._is_at_end():
            self.current += 1
        return self._previous()

    def _is_at_end(self) -> bool:
        """Check if at end of tokens."""
        return self._peek().type == TokenType.EOF

    def _peek(self) -> Token:
        """Peek at current token."""
        return self.tokens[self.current]

    def _previous(self) -> Token:
        """Get previous token."""
        return self.tokens[self.current - 1]
